xarray/core/interp.py

```diff
@@ -0,0 +1,152 @@
+from __future__ import absolute_import, division, print_function
+from functools import partial
+
+import numpy as np
+from .computation import apply_ufunc
+from .pycompat import dask_array_type
+from .variable import broadcast_variables
+
+
+def _localize(obj, index_coord):
+    """ Speed up for linear and nearest neighbor method.
+    Only consider a subspace that is needed for the interpolation
+    """
+    for dim, [x, new_x] in index_coord.items():
+        try:
+            imin = x.to_index().get_loc(np.min(new_x), method='ffill')
+            imax = x.to_index().get_loc(np.max(new_x), method='bfill')
+
+            idx = slice(np.maximum(imin - 1, 0), imax + 1)
+            index_coord[dim] = (x[idx], new_x)
+            obj = obj.isel(**{dim: idx})
```

Comment:
Currently, only a small portion of arrays will be used for the interpolation with `method='linear' / 'nearest '`. So if there are no NaN in this region, we can interpolate this array.